src/server/server_impl.go

```diff
@@ -73,11 +76,20 @@ func (server *server) AddJsonHandler(svc pb.RateLimitServiceServer) {
 			http.Error(writer, err.Error(), http.StatusBadRequest)
 			return
 		}
+
 		logger.Debugf("resp:%s", resp)
+
+		writer.Header().Set("Content-Type", "application/json")
 		if resp.OverallCode == pb.RateLimitResponse_OVER_LIMIT {
-			http.Error(writer, "over limit", http.StatusTooManyRequests)
+			writer.WriteHeader(http.StatusTooManyRequests)
 		} else if resp.OverallCode == pb.RateLimitResponse_UNKNOWN {
-			http.Error(writer, "unknown", http.StatusInternalServerError)
+			writer.WriteHeader(http.StatusInternalServerError)
+		}
+
+		err = m.Marshal(writer, resp)
+		if err != nil {
```

Comment:
Can this actually happen? If not can you assert it instead?

Reply:
My _guess_ is that there probably couldn't be a serialization error but probably could be an I/O error if the remote peer closes the TCP connection, in which case the fmt.Fprintf() would fail anyway.

I could split the serialization from the writing to the socket and do a assert/panic and recover dance to print the stack trace if a serialization error happens. If that's the route, I'd kind of rather just do a normal log.Error() instead of an assert because I don't think the extra code to deal with panic()s to get a stack trace will provide much value.

Up to you.

Reply:
I'm fine either way, I mostly just want to avoid error handling and just crash if there are things that should never happen.

Reply:
My sense is:

- handling the error differently than how it's handled here is probably more code
- I'd be comfortable leaving the code here as-is and waiting until someone has evidence that a json serialization error can actually happen before writing additional logic to handle it in a more nuanced way

More generally on the topic of when to crash the process, it seems like crashing a process makes sense when:
- The process is permanently severely degraded or broken
- Restarting has a decent chance of un-wedging it

My sense is that a single unknown json serialization error in isolation probably doesn't justify a crash, vs something like a resource allocation issue potentially could justify a crash.

Reply:
The guiding principle is whether you can test this or not. I don't think you can test this branch, as such I would use https://github.com/envoyproxy/ratelimit/blob/master/src/assert/assert.go and not handle the error.

Reply:
Added unit tests for all the cases. For the serialization failure I used a 'nil' return value from ShouldRateLimit(). A non-nil serialization problem would theoretically be possible if the ShouldRateLimit() API ever evolves to include `Any` types, but I don't think there are `Any` types in the spec today

Reply:
Sounds good, thanks. Will take a look.